cep-20.1.md

```diff
@@ -0,0 +1,352 @@
+# A new recipe format – part 1
+
+<table>
+<tr><td> Title </td><td> A new recipe format </td>
+<tr><td> Status </td><td> Proposed</td></tr>
+<tr><td> Author(s) </td><td> Wolf Vollprecht &lt;wolf@prefix.dev&gt;</td></tr>
+<tr><td> Created </td><td> May 23, 2023</td></tr>
+<tr><td> Updated </td><td> May 23, 2023</td></tr>
+<tr><td> Discussion </td><td>  </td></tr>
+<tr><td> Implementation </td><td>https://github.com/prefix-dev/rattler-build</td></tr>
+</table>
+
+## Abstract
+
+We propose a new recipe format that is heavily inspired by conda-build. The main
+change is a pure YAML format without arbitrary Jinja or comments with semantic
+meaning.
+
+## Motivation
+
+The conda-build format has grown over the years to become quite complex.
+Unfortunately it has never been formally "specified" and it has grown some
+features over time that make it hard to parse as straightforward YAML.
+
+The CEP attempts to introduce a subset of the conda build format that allows for
+fast parsing and building of recipes.
+
+### History
+
+A discussion was started on what a new recipe spec could or should look like.
+The fragments of this discussion can be found here:
+https://github.com/mamba-org/conda-specs/blob/master/proposed_specs/recipe.md
+The reason for a new spec are:
+
+- Make it easier to parse ("pure yaml"). conda-build uses a mix of comments and
+  jinja to achieve a great deal of flexibility, but it's hard to parse the
+  recipe with a computer
+- iron out some inconsistencies around multiple outputs (build vs. build/script
+  and more)
+- remove any need for recursive parsing & solving
+- cater to needs for automation and dependency tree analysis via a determinstic
+  format
+
+## Major differences with conda-build
+
+- no full Jinja2 support: no block support `{% set ...` support, only string
+  interpolation. Variables can be set in the toplevel "context" which is valid
+  YAML (all new features should be native to YAML specs)
+- Jinja variable syntax is changed to begin with `${{` so that it becomes valid
+  YAML, e.g. `- ${{ version }}`
+- Selectors use a YAML dictionary with `if / then / else` (vs. comments in
+  conda-build) and are only allowed in lists (dependencies, scripts, ...). The
+  syntax looks like:
+  ```yaml
+  - if: win
+    then: this
+    else: that  # optional
+  ```
+- for inline values, the Jinja ternary operator can be used, e.g. `number: ${{ 0
```

Comment:
How crazy can these get? For example, are these ok?

- `${{ 0 if (linux or osx) else 1 }}`
- `${{ 0 if (linux or osx) else (100 if cuda else 50) }}`

What about defining new blocks in a Jinja expression?

```yaml
requirements:
  build:
    - ${{ compiler('cxx') }}
    - ${{ {"if": "unix", "then": "make"} if python_version >= (3,10) else None }}
```

Reply:
Yes, nesting if statements works. You can make multiline strings and have crazy expressions if you want, e.g.

```yaml
- |
  ${{ "bla" if python_version >= (3,10) else 
      (100 if other else 
      (200 if that else "123123"
    }}
```

Minijinja has a nice online playground, btw: https://mitsuhiko.github.io/minijinja-playground/

We can enforce that only `bool, int, string` are allowed evaluation types of minijinja expressions (lists, dicts and floats would be forbidden).